uber/models/core.py

```diff
@@ -0,0 +1,1094 @@
+            return leader.email
+        else:
+            emails = [a.email for a in self.attendees if a.email]
+            if len(emails) == 1:
+                return emails[0]
+
+    @property
+    def badges_purchased(self):
+        return len([a for a in self.attendees if a.paid == c.PAID_BY_GROUP])
+
+    @property
+    def badges(self):
+        return len(self.attendees)
+
+    @property
+    def unregistered_badges(self):
+        return len([a for a in self.attendees if a.is_unassigned])
+
+    @cost_property
+    def table_cost(self):
+        table_count = int(float(self.tables))
+        return sum(c.TABLE_PRICES[i] for i in range(1, 1 + table_count))
+
+    @property
+    def new_badge_cost(self):
+        return c.DEALER_BADGE_PRICE if self.is_dealer else c.get_group_price()
+
+    @cost_property
+    def badge_cost(self):
+        total = 0
+        for attendee in self.attendees:
+            if attendee.paid == c.PAID_BY_GROUP:
+                total += attendee.badge_cost
+        return total
+
+    @property
+    def amount_extra(self):
+        if self.is_new:
+            return sum(
+                a.total_cost - a.badge_cost for a in self.attendees
+                if a.paid == c.PAID_BY_GROUP)
+        else:
+            return 0
+
+    @property
+    def total_cost(self):
+        return self.default_cost + self.amount_extra
+
+    @property
+    def amount_unpaid(self):
+        if self.registered:
+            return max(0, self.cost - self.amount_paid)
+        else:
+            return self.total_cost
+
+    @property
+    def dealer_max_badges(self):
+        return math.ceil(self.tables) + 1
+
+    @property
+    def dealer_badges_remaining(self):
+        return self.dealer_max_badges - self.badges
+
+    @property
+    def hours_since_registered(self):
+        if not self.registered:
+            return 0
+        delta = datetime.now(UTC) - self.registered
+        return max(0, delta.total_seconds()) / 60.0 / 60.0
+
+    @property
+    def hours_remaining_in_grace_period(self):
+        return max(
+            0, c.GROUP_UPDATE_GRACE_PERIOD - self.hours_since_registered)
+
+    @property
+    def is_in_grace_period(self):
+        return self.hours_remaining_in_grace_period > 0
+
+    @property
+    def min_badges_addable(self):
+        if self.can_add:
+            return 1
+        elif self.is_dealer:
+            return 0
+        else:
+            return c.MIN_GROUP_ADDITION
+
+    @property
+    def requested_hotel_info(self):
+        if self.leader:
+            return self.leader.requested_hotel_info
+        elif self.leader_id:  # unattached groups
+            for attendee in self.attendees:
+                if attendee.id == self.leader_id:
+                    return attendee.requested_hotel_info
+        else:
+            return any(a.requested_hotel_info for a in self.attendees)
+
+
+class Attendee(MagModel, TakesPaymentMixin):
```

Comment:
Y'know... maybe the Attendee class should have its own file... 😰

Reply:
Not a bad idea. The `Attendee` class is huge!